### PR TITLE
multiline processing for tail input plugin

### PR DESCRIPTION
**Pull request by javicrespo**
Implements #1818/#3228

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by 0UserName** [comment]
When you plan to merge this PR?

**Comment by lsitzmann** [comment]
Hi.
I have a short question about this PR. In the docs you reference https://www.elastic.co/guide/en/logstash/2.4/plugins-filters-multiline.html and there they say that this multiline-implementation isn't thread-safe.
Does this also apply for this implementation in telegraf? Is it safely possible to tail multiple streams with multiline-support using this implementation?

Thanks, Lars

**Comment by javicrespo** [comment]
@lsitzmann I referenced the logstash doc because the configuration options are a heavily inspired (or rather a copy) by the logstash plugin.
With regards to thread safety, the tailer implementation is based on goroutines so it should be thread safe (take this with a grain of salt cause I'm nowhere near being a Golang expert).
Each file is individually tailed and the multi line buffering is also managed in a separate and dedicated golang channel.

- One "tailer" per file (stream?)
https://github.com/influxdata/telegraf/blob/839ceed9867cdd78a0ebfbb5b000885648a70452/plugins/inputs/tail/tail.go#L135

- One golang channel per tailer (contains the multine buffer handling)
https://github.com/influxdata/telegraf/blob/839ceed9867cdd78a0ebfbb5b000885648a70452/plugins/inputs/tail/tail.go#L169

**Comment by lsitzmann** [comment]
@danielnelson : Are there any chances, that this PR will be merged soon? Currently we have to manually copy log files from several production servers because the data collected with telegraf isn't complete.
We are really looking foreward for this feature.

**Comment by lijingwei9060** [comment]
@javicrespo  , i think channelOpen(telegraf/plugins/inputs/tail/tail.go line 212) should be initiated with true, otherwise i will get "Error in plugin: E! Error tailing file /var/log/bingocloud/cc2.log, Error: tomb: still alive"

**Comment by javicrespo** [comment]
@lijingwei9060  Should be fixed with the last commit. I've also rewritten the timeout unit test to repro the bug.

**Comment by lsitzmann** [comment]
Hi again.
We are using this PR in a production environment for about one month now. We parsed about 500k lines of logfiles with it and it seems to work very reliably - at least in our configuration.
About 1% of the log-entries are multi-lines.

**We'd really appreciate to switch back to an officially released version of telegraf** so once again my question: is it foreseeable  when this PR will make it into an release??

**Comment by danielnelson** [comment]
I've added this to the 1.12 milestone, sorry I know it is frustrating to wait for pull requests to be merged but I feel code review is required in order to keep our bug count at a reasonable level.

File: plugins/inputs/tail/tail.go [hunk 1]
```diff
@@ -64,6 +69,22 @@ const sampleConfig = `
   ## more about them here:
   ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
   data_format = "influx"
+
+  ## multiline parser/codec
```
Comment (glinton) [inline plugins/inputs/tail/tail.go]: Make this match the tail readme

File: plugins/inputs/tail/tail.go [hunk 2]
```diff
@@ -202,6 +251,45 @@ func (t *Tail) receiver(parser parsers.Parser, tailer *tail.Tail) {
...
+			if len(metrics) == 0 {
+				return
+			}
+			m = metrics[0]
+		}
+	} else {
+		m, err = parser.ParseLine(text)
+	}
+
+	if err == nil {
```
Comment (glinton) [inline plugins/inputs/tail/tail.go]: Flip this conditional and return after`AddError`

File: plugins/inputs/tail/tail.go [hunk 3]
```diff
@@ -202,6 +251,45 @@ func (t *Tail) receiver(parser parsers.Parser, tailer *tail.Tail) {
 	}
 }
 
+func (t *Tail) getLineText(line *tail.Line, filename string) string {
+	if line.Err != nil {
+		t.acc.AddError(fmt.Errorf("E! Error tailing file %s, Error: %s\n",
```
Comment (glinton) [inline plugins/inputs/tail/tail.go]: Add `[inputs.tail]` to this and other errors printed/added

File: plugins/inputs/tail/testdata/test-patterns [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Test multiline
+#   [04/Jun/2016:12:41:45 +0100] DEBUG HelloExample: This is debug
+TEST_LOG_MULTILINE \[%{HTTPDATE:timestamp:ts-httpd}\] %{WORD:loglevel:tag} %{GREEDYDATA:message}
```
Comment (glinton) [inline plugins/inputs/tail/testdata/test-patterns]: Nit: add newline at end of this file

**Comment by dsinang** [comment]
Hello, would really like to see this PR merged and released.

Any idea on when that may be ?

**Comment by DSpeichert** [comment]
@javicrespo I think this PR needs rebasing at this point.

**Comment by zibuyu1995** [comment]
this PR will make it into an release?

**Comment by ssoroka** [comment]
waiting on @javicrespo to implement on top of the latest master (the merge from master might have confused things there). If not interested, we or someone else could implement this feature.

**Comment by javicrespo** [comment]
Considering that I don’t need this feature anymore and the effort to retrofit it to the latest version in master seems non-trivial to me, I’m going to pass. Feel free to pick it up 🙂

**Comment by lsitzmann** [comment]
We‘d really appreciate if anyone would do the integration of this feature. 
We use it on a lot of systems- and now we are stuck to and old and self-build version of telegraf.

**Comment by ssoroka** [comment]
Closing this in favor of https://github.com/influxdata/telegraf/pull/7309

**Pull request closed without merging**